### DISCUSSION TITLE
--tomogram-mask option during match_template not working

### DISCUSSION THREAD

**Discussion by Phaips** [SBC-Utrecht/pytom-match-pick#211] [Questions]
Hi again,
When I look at my score maps I see peaks inside the area where my tomogram mask should have zero values. I assume this shouldn't be the case if those subvolumes are skipped. 

My mask is from Amira and has the right dimensions. However, the pixel size is 1.00. Now I was wondering if the pixel size of the mask needs to match the tomogram? 
Or there is another problem at play :)

Thanks a lot!

**Comment by sroet** [comment]
Dear @Phaips 

First to check:
>My mask is from Amira and has the right dimensions. 

Do you mean it has the right size in angstrom or in pixels? We need it to be **pixels**, but I expect our code to fail if the number of pixels isn't identical, at least it *should* (not sure if it actually does, and this might be the bug you're encountering)

> When I look at my score maps I see peaks inside the area where my tomogram mask should have zero values. I assume this shouldn't be the case if those subvolumes are skipped.

Another reason for this behavior might be:
Our current implementation only skips subvolumes where **all** values are masked (with the `--volume-split` option), and we won't zero out data in *partially* masked subvolumes (to not throw away potential useful data we loop over anyway). 

During the extract particles step, these particles **should** be skipped automatically

**Comment by Phaips** [comment]
Hi @sroet,
Thanks for the quick reply! Yes with dimensions I meant pixels, so bin4 I have 1024,1024,512 x,y,z for both my mask and tomogram respectively. Only size per pixel is different (7.84 A/px for the tomo and 1 A/px for the mask) which I guess shouldn't matter?

> Our current implementation only skips subvolumes where all values are masked (with the --volume-split option), and we won't zero out data in partially masked subvolumes (to not throw away potential useful data we loop over anyway).

I am not sure I understood correctly but I think this explains the observation! I do use `--volume-split 2 2 1` and so this means it will not zero out the data if not one of the entire 4 subvolumes is masked out completely? But it will then skip those without supplying the tomo mask again in the extraction step?
Sorry for my confusion :D

Cheers!

**Comment by sroet** [comment]
> I am not sure I understood correctly but I think this explains the observation! I do use --volume-split 2 2 1 and so this means it will not zero out the data if not one of the entire 4 subvolumes is masked out completely? But it will then skip those without supplying the tomo mask again in the extraction step?
Sorry for my confusion :D

No worries about the confusion, but you are absolutely correct. With a 2 2 1 split it will only skip a subvolume if one of the 4 subvolumes is completely masked. (This was mainly done to make sure the unmasked results were not altered by supplying a mask). We could set the masked scores to 0 at the last step, but throwing away data at that point felt wasteful.

It should indeed automatically grab the mask from the matching job in the extraction job. If you don't want that you can supply a different mask explicitly to the extraction job 

Please let me know if anything is unclear!

**Comment by McHaillet** [comment]
Did we need to make a PR for this? Otherwise it would be nice to move to Discussions for public reference

**Comment by sroet** [comment]
PR (#210) has been made to make sure we error with a reasonable error out if the shape does not overlap.
Moving this to discussions instead
